src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py

```diff
@@ -96,3 +231,70 @@ def list_snapshots(self):
             if ve.errno == -errno.ENOENT:
                 return []
             raise
+
+    def _protect_snapshot(self, snapname):
+        try:
+            self.metadata_mgr.add_section("protected snaps")
+            self.metadata_mgr.update_section("protected snaps", snapname, "1")
+            self.metadata_mgr.flush()
+        except MetadataMgrException as me:
+            log.warn("error updating protected snap list ({0})".format(me))
+            raise VolumeException(-errno.EINVAL, "error protecting snapshot")
+
+    def _unprotect_snapshot(self, snapname):
+        try:
+            self.metadata_mgr.remove_option("protected snaps", snapname)
+            self.metadata_mgr.flush()
+        except MetadataMgrException as me:
+            log.warn("error updating protected snap list ({0})".format(me))
+            raise VolumeException(-errno.EINVAL, "error unprotecting snapshot")
+
+    def protect_snapshot(self, snapname):
+        if not snapname.encode('utf-8') in self.list_snapshots():
+            raise VolumeException(-errno.ENOENT, "snapshot '{0}' does not exist".format(snapname))
+        if self.is_snapshot_protected(snapname):
+            raise VolumeException(-errno.EEXIST, "snapshot '{0}' is already protected".format(snapname))
+        self._protect_snapshot(snapname)
+
+    def unprotect_snapshot(self, snapname):
+        if not snapname.encode('utf-8') in self.list_snapshots():
+            raise VolumeException(-errno.ENOENT, "snapshot '{0}' does not exist".format(snapname))
+        if not self.is_snapshot_protected(snapname):
+            raise VolumeException(-errno.EEXIST, "snapshot '{0}' is not protected".format(snapname))
+        if self.has_pending_clones(snapname):
```

Comment:
can you unprotect snapshot if the clone has failed?

Reply:
ditto